src/field/fieldperp.cxx

```diff
@@ -48,6 +48,16 @@ FieldPerp::FieldPerp(BoutReal val, Mesh *localmesh) : FieldPerp(localmesh) {
   *this = val;
 }
 
+FieldPerp::FieldPerp(Array<BoutReal> data_in, Mesh* localmesh, CELL_LOC location_in,
+                     int yindex_in, DirectionTypes directions)
+    : Field(localmesh, location_in, directions), yindex(yindex_in),
+      nx(fieldmesh->LocalNx), nz(fieldmesh->LocalNz), data(std::move(data_in)) {
+
+  ASSERT1(data.size() == nx * nz);
+
+  setLocation(location_in);
```

Comment:
Is this call necessary? If `setLocation()` is doing anything that's needed, shouldn't it be in the `Field()` constructor rather than here?

Reply:
Good point! I copied this from the equivalent `Field3D` constructor. `setLocation` is non-trivial, and normalises/checks `location` (`CELL_DEFAULT` -> `CELL_CENTRE`, throws if `CELL_VSHIFT`, etc). You're right that this is better done in the `Field` constructor though